File: (结构)享元模式.v2/golang/gpt/index.享元.go
package main

import "fmt"

type Flyweight interface {
	Operation(extrinsicState string)
}

type ConcreteFlyweight struct {
	intrinsicState string // 内部状态
}

func (c *ConcreteFlyweight) Operation(extrinsicState string) {
	fmt.Printf("ConcreteFlyweight: Intrinsic State %s, "+
		"Extrinsic State %s\n", c.intrinsicState, extrinsicState)
}

type FlyweightFactory struct {
	flyweights map[string]Flyweight
}

func NewFlyweightFactory() *FlyweightFactory {
	return &FlyweightFactory{
		flyweights: make(map[string]Flyweight),
	}
}

func (f *FlyweightFactory) GetFlyweight(key string) Flyweight {
	if flyweight, ok := f.flyweights[key]; ok {
		return flyweight
	}
	flyweight := &ConcreteFlyweight{intrinsicState: key}
	f.flyweights[key] = flyweight
	return flyweight
}

func main() {
	factory := NewFlyweightFactory()

	flyweight1 := factory.GetFlyweight("foo")
	flyweight1.Operation("bar1")

	flyweight2 := factory.GetFlyweight("foo") // 它应该返回相同的对象
	flyweight2.Operation("bar2")

	fmt.Printf("flyweight1 == flyweight2: %v\n", flyweight1 == flyweight2) // 输出 true, 显示它们是相同的实例
}

File: (结构)享元模式.v2/php/gpt/index.享元.php
<?php
namespace php\gpt\fly_weight;

interface FlyweightInterface {
    public function render($extrinsicState);
}

class ConcreteFlyweight implements FlyweightInterface {
    private $intrinsicState;

    public function __construct($intrinsicState) {
        $this->intrinsicState = $intrinsicState;
    }

    public function render($extrinsicState) {
        // 实际中，根据内部状态和外部状态完成某些功能，例如渲染
        return sprintf('Intrinsic State: %s, Extrinsic State: %s', $this->intrinsicState, $extrinsicState);
    }
}

class FlyweightFactory {
    private $pool = [];

    public function getFlyweight($intrinsicState) {
        if (!isset($this->pool[$intrinsicState])) {
            $this->pool[$intrinsicState] = new ConcreteFlyweight($intrinsicState);
        }

        return $this->pool[$intrinsicState];
    }
}

$flyweightFactory = new FlyweightFactory();

$flyweightA = $flyweightFactory->getFlyweight('stateA');
echo $flyweightA->render('extraState1');
echo "\n";
$flyweightB = $flyweightFactory->getFlyweight('stateA');
echo $flyweightB->render('extraState2');

// 由于stateA的内部状态相同，$flyweightA 和 $flyweightB 指向的是同一个实例

//Intrinsic State: stateA, Extrinsic State: extraState1
//Intrinsic State: stateA, Extrinsic State: extraState2

